katdal/test/test_applycal.py

```diff
@@ -155,36 +174,52 @@ def bandpass_corrections(pol, ant):
     """Figure out N_DUMPS gain corrections given `pol` and `ant` indices."""
     dumps = np.arange(N_DUMPS)
     events = np.array(GAIN_EVENTS)
-    gains = create_gain(pol, ant)
-    valid = np.isfinite(gains)
-    if valid.any():
-        gains = complex_interp(dumps, events[valid], gains[valid])
+    gains = create_gain(pol, ant, multi_channel, targets)
+    gains = np.atleast_2d(gains.T)
+    if targets:
+        targets = CategoricalData(np.arange(len(GAIN_EVENTS)) % 2,
+                                  GAIN_EVENTS + [N_DUMPS])
     else:
-        gains = np.full(N_DUMPS, INVALID_GAIN)
-    return np.reciprocal(gains)
-
-
-def corrections_per_corrprod(dumps, channels, corrprods=()):
+        targets = CategoricalData([0], [0, len(dumps)])
+    smooth_gains = np.empty((N_DUMPS, gains.shape[0]), dtype=gains.dtype)
+    for chan, gains_per_chan in enumerate(gains):
+        for target in set(targets):
+            on_target = (targets == target)
+            valid = np.isfinite(gains_per_chan) & on_target[events]
+            smooth_gains[on_target, chan] = INVALID_GAIN if not valid.any() else \
+                complex_interp(dumps[on_target], events[valid], gains_per_chan[valid])
+    return np.reciprocal(smooth_gains)
+
+
+def corrections_per_corrprod(dumps, channels, cal_products):
     """Predict corrprod correction for a time-frequency-baseline selection."""
     input_map = {ant + pol: (pol_idx, ant_idx)
                  for (pol_idx, pol) in enumerate(POLS)
                  for (ant_idx, ant) in enumerate(ANTS)}
     gains_per_input = np.ones((len(dumps), N_CHANS, len(INPUTS)),
                               dtype='complex64')
-    # Apply (K, B, G) corrections
-    gains_per_input *= np.array([delay_corrections(*input_map[inp])
-                                 for inp in INPUTS]).T
-    gains_per_input *= np.array([bandpass_corrections(*input_map[inp])
-                                 for inp in INPUTS]).T
-    gains_per_input *= np.array([gain_corrections(*input_map[inp])[dumps]
-                                 for inp in INPUTS]).T[:, np.newaxis]
+    corrections = {
+        CAL_STREAM + '.K': np.array([delay_corrections(*input_map[inp])
+                                     for inp in INPUTS]).T,
+        CAL_STREAM + '.B': np.array([bandpass_corrections(*input_map[inp])
+                                     for inp in INPUTS]).T,
+        CAL_STREAM + '.G': np.array([gain_corrections(*input_map[inp])[dumps]
+                                     for inp in INPUTS]).T,
+        CAL_STREAM + '.GPHASE': np.array([gain_corrections(*input_map[inp],
+                                                           multi_channel=True,
+                                                           targets=True)[dumps]
+                                          for inp in INPUTS]
+                                         ).transpose(1, 2, 0)[:, DATA_TO_CAL_CHANNEL]}
+    # Apply (K, B, G, GPHASE) corrections in correct order
```

Comment:
What does "correct" order mean? Is this where you just want to match the order used by the system under test to get bit-exact results?

Reply:
Exactly.